crypto/x509/x509_cmp.c

```diff
@@ -440,19 +440,23 @@ int X509_CRL_check_suiteb(X509_CRL *crl, EVP_PKEY *pk, unsigned long flags)
 }
 
 #endif
+
+static X509 *x509upref(X509 *x)
+{
+    if (!X509_up_ref(x))
+        return NULL;
+
+    return x;
+}
+
 /*
  * Not strictly speaking an "up_ref" as a STACK doesn't have a reference
  * count but it has the same effect by duping the STACK and upping the ref of
  * each X509 structure.
  */
 STACK_OF(X509) *X509_chain_up_ref(STACK_OF(X509) *chain)
 {
-    STACK_OF(X509) *ret;
-    int i;
-    ret = sk_X509_dup(chain);
-    for (i = 0; i < sk_X509_num(ret); i++) {
-        X509 *x = sk_X509_value(ret, i);
-        X509_up_ref(x);
-    }
-    return ret;
+    /* We upref each element rather than actually copying */
+    return sk_X509_deep_copy(chain, (X509 *(*)(const X509 *))x509upref,
```

Comment:
I'm also not a big fan of abusing the deep_copy function this way. It confuses the reader when they're reading the code thinking its doing a deep copy and not just a shallow copy like the original code implied.

Reply:
Arguably the sk_X509_dup function is misnamed as a shallow-duplicated stack should involve bumping the reference counts of all elements automatically.

Reply:
One might argue that `sk_*_deep_copy` is misnamed, and indicates an intended use rather than the actual functionality.  Basically, the "depth" of the copy is governed by the supplied copyfunc rather than by `sk_*_deep_copy` itself.

The right thing to do would probably be to rename `sk_*_deep_copy` to just `sk_*_copy`, and provide the former name as a (deprecated?) alias.

Reply:
That would make sense, but this needs to be backported to 1.1.1.

Reply:
That's fine.  I think we can live with this "misuse" of `sk_*_deep_copy`...  but perhaps a comment that remarks on this would be suitable.

Reply:
Agreed, a comment would be beneficial.